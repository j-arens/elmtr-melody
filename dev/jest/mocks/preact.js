module.exports = {
    unmountComponent: jest.fn(),
};
