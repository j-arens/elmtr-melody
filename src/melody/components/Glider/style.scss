.glider {
    position: relative;
    margin-bottom: -2rem;

    &__playback {
        display: flex;
        align-items: center;
        height: 16px;
        padding: 0 14px 0 0;
        overflow: hidden;
    }

    &__trackbar {
        position: relative;
        background-color: $color_white;
        margin-left: 0;
        background-color: transparent;
        
        &--transition {
            transition: margin 1s linear;
        }

        &::before,
        &::after {
            content: '';
            position: absolute;
            top: 50%;
            transform: translateY(-50%);
            height: 5px;
            width: 9999px;
        }
        
        &::before {
            right: 0;
        }
        
        &::after {
            left: 100%;
        }
    }

    &__handle {
        position: relative;
        z-index: 1;
        padding: 7px;
        border-radius: 50%;
        box-shadow: 2px 0px 2px 1px rgba(0, 0, 0, 0.2);
    }
}