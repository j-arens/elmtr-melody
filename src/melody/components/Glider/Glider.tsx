import DragHelper from '@components/DragHelper/';
import { DragProps, MelodyDragEvent } from '@components/DragHelper/type';
import { ELEMENTOR_NO_DRAG } from '@constants';
import { Action } from '@redux/type';
import { MachineStates } from '@state-machine/type';
import { getLayerX } from '@utils/index';
import { NO_OP } from '@utils/index';
import * as classnames from 'classnames';
import { Component, h } from 'preact';
const s = require('./style.scss');

interface Props extends DragProps {
    currentState: MachineStates;
    duration: number;
    currentTime: number;
    updateCurrentTime: (updatedTime: number) => Action;
    toggleGliderDragging: () => Action;
    triggerTimeSync: () => Action;
}

interface State {
    transitionHandle: boolean;
    offset: number;
    clientX: number;
}

const HANDLE_AREA: number = 14;
const HANDLE_SPACING: number = 2;

export default class extends Component<Props, State> {
    state = {
        transitionHandle: false,
        offset: 0,
        clientX: 0,
    };

    ref: HTMLElement;

    componentDidMount() {
        const { onDragStart, onDrag, onDragEnd } = this.props;
        onDragStart(this.onDragStart);
        onDrag(this.onDrag);
        onDragEnd(this.onDragEnd);
    }

    componentDidUpdate(prevProps: Props) {
        if (this.props.currentTime !== prevProps.currentTime) {
            const { duration, currentTime } = this.props;
            if (duration) {
                const nextOffset: number = Math.round(100 / (duration / currentTime));
                this.updateOffset(nextOffset);
            } else {
                this.updateOffset(0);
            }
        }

        if (this.props.isDragging !== prevProps.isDragging) {
            this.props.toggleGliderDragging();
        }
    }

    setRef = (ref: HTMLElement) => {
        this.ref = ref;
        this.props.setDragRef(ref);
    }

    eventGuard = fn => (...args) => {
        if (!this.props.duration) {
            return NO_OP;
        }
        return fn(...args);
    }

    updateOffset(nextOffset: number) {
        this.setState({ offset: nextOffset });
    }

    onDragStart = this.eventGuard(({ clientX }: MelodyDragEvent) => {
        this.setState({ clientX });
    });

    onDragEnd = this.eventGuard((e: MelodyDragEvent) => {
        this.handleClick(e, false);
    });

    onDrag = this.eventGuard(({ clientX }: MelodyDragEvent) => {
        if (!this.ref) {
            return;
        }

        const originalClientX = this.state.clientX;

        if (clientX === originalClientX) {
            return;
        }

        const { offset } = this.state;
        const { duration, updateCurrentTime } = this.props;
        const width: number = this.ref.offsetWidth;
        const x: number = getLayerX(this.ref, clientX);
        let newOffset: number = Number(((x / width) * 100).toFixed(2));

        if (newOffset > 100) {
            newOffset = 100;
        }

        if (newOffset < 0) {
            newOffset = 0;
        }

        this.updateOffset(newOffset);
        const updatedTime: number = Number(((duration * newOffset) / 100).toFixed(0));
        updateCurrentTime(updatedTime);
    });

    handleClick = this.eventGuard(({ clientX }: MouseEvent, ignoreHandle: boolean = true) => {
        if (!this.ref) {
            return;
        }

        const width: number = this.ref.offsetWidth;
        const x: number = getLayerX(this.ref, clientX);

        if (ignoreHandle) {
            const currentOffset: number = this.state.offset;
            const rightPadding: number = HANDLE_AREA / 2;
            const handleCenter: number = (currentOffset * width) / 100;
            const lowerBoundary: number = handleCenter - (HANDLE_AREA + HANDLE_SPACING) + rightPadding;
            const upperBoundary: number = handleCenter + HANDLE_AREA + HANDLE_SPACING;
            const withinHandle: boolean = x >= lowerBoundary && x <= upperBoundary;

            if (withinHandle) {
                return;
            }
        }

        const { duration, updateCurrentTime, triggerTimeSync } = this.props;
        const xPercent: number = 100 / (width / x);
        const updatedTime: number = Number(((duration * xPercent) / 100).toFixed(0));
        updateCurrentTime(updatedTime);
        triggerTimeSync();
    });

    render({ setDragRef, currentState, isDragging }, { offset }) {
        const defaultClass = s.glider__trackbar;
        const shouldTransition = currentState === 'playing' && offset > 0 && !isDragging;
        const classes = classnames(defaultClass, {
            [`${defaultClass}--transition`]: shouldTransition,
        });
        return (
            <div
                ref={(ref: HTMLElement) => this.setRef(ref)}
                class={`${s.glider} ${ELEMENTOR_NO_DRAG}`}
                onMouseDown={this.handleClick}
            >
                <div class={s.glider__playback}>
                    <div
                        class={classes}
                        style={{marginLeft: `${offset}%`}}
                        data-melody-scrubber-progress
                    >
                        <div class={s.glider__handle} data-melody-scrubber-handle />
                    </div>
                </div>
            </div>
        );
    }
}
