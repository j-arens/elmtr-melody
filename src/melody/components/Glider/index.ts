import { triggerTimeSync, updateCurrentTime } from '@redux/modules/audio/actions';
import { toggleGliderDragging } from '@redux/modules/ui/actions';
import { State } from '@redux/type';
import { connect } from 'preact-redux';
import DragHelper from '../DragHelper';
import Glider from './Glider';

const mapStateToProps = (state: State) => ({
    currentState: state.state.currentState,
    currentTime: state.audio.currentTime,
    duration: state.audio.filelength,
});

const mapDispatchToProps = (dispatch) => ({
    updateCurrentTime: (updatedTime: number) => dispatch(updateCurrentTime(updatedTime)),
    toggleGliderDragging: () => dispatch(toggleGliderDragging()),
    triggerTimeSync: () => dispatch(triggerTimeSync()),
});

export default connect(mapStateToProps, mapDispatchToProps)(DragHelper(Glider));
