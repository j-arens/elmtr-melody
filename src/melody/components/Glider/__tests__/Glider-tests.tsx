import { render } from 'preact-render-to-string';
import { h } from 'preact';
import { NO_OP } from '@utils';
import Glider from '../Glider';

describe('Glider', () => {
    it('renders', () => {
        const component = render(
            <Glider
                currentState="stopped"
                duration={100}
                currentTime={0}
                updateCurrentTime={NO_OP}
                toggleGliderDragging={NO_OP}
                triggerTimeSync={NO_OP}
            />
        );
        expect(component).toMatchSnapshot();
    });
});
