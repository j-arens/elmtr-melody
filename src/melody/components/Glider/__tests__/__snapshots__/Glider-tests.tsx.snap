// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Glider renders 1`] = `
<div class="glider elementor-tab-title">
  <div class="glider__playback">
    <div class="glider__trackbar" style="margin-left: 0%;" data-melody-scrubber-progress>
      <div class="glider__handle" data-melody-scrubber-handle data-cy="glider-handle"></div>
    </div>
  </div>
</div>
`;
